client/pkg/omnictl/config/config.go

```diff
@@ -194,11 +200,112 @@ func (c *Config) Merge(additionalConfigPath string) ([]Rename, error) {
 	return renames, nil
 }
 
-func defaultPath() (string, error) {
+type accessType int32
+
+const (
+	READ accessType = iota
+	WRITE
+)
+
+//nolint:gocognit
+func defaultPath(access accessType) (string, error) {
 	path := os.Getenv(OmniConfigEnvVar)
 	if path != "" {
 		return path, nil
 	}
 
-	return xdg.ConfigFile(relativePath)
+	baseDir, err := config.GetTalosDirectory()
+	if err == nil {
+		if access == WRITE {
+			if !fileutils.FileExists(filepath.Join(baseDir, OmniRelativePath)) && fileutils.FileExists(filepath.Join(xdg.ConfigHome, OmniRelativePath)) {
+				// Ensure the Talos home directory exists and is writable.
+				if _, err = ensurePath(filepath.Join(baseDir, omniConfigFolder)); err == nil {
+					if fileutils.IsWritable(filepath.Join(baseDir, omniConfigFolder)) {
+						// Attempt to copy the config from XDG config directory to Talos home directory when all the dependencies are satisfied.
+						// This normally shouldn't fail, but if it does, do not fall back to XDG config directory, and fail with error.
+						return copyConfig(filepath.Join(xdg.ConfigHome, OmniRelativePath), filepath.Join(baseDir, omniConfigFolder))
+					}
+				}
+
+				// Do not try to fall back to XDG config directory for WRITE access
+				return "", err
+			}
+
+			return ensurePath(filepath.Join(baseDir, omniConfigFolder))
+		}
+
+		if access == READ {
+			if !fileutils.FileExists(filepath.Join(baseDir, OmniRelativePath)) && fileutils.FileExists(filepath.Join(xdg.ConfigHome, OmniRelativePath)) {
+				// Ensure the Talos home directory exists and is writable.
+				if _, err = ensurePath(filepath.Join(baseDir, omniConfigFolder)); err == nil {
+					if fileutils.IsWritable(filepath.Join(baseDir, omniConfigFolder)) {
+						// Attempt to copy the config from XDG config directory to Talos home directory when all the dependencies are satisfied.
+						// This normally shouldn't fail, but if it does, do not fall back to XDG config directory, and fail with error.
+						return copyConfig(filepath.Join(xdg.ConfigHome, OmniRelativePath), filepath.Join(baseDir, omniConfigFolder))
+					}
+				}
+
+				fmt.Fprintf(os.Stderr, "WARN: Default Omni config location: '%s' is not writable, falling back to deprecated Omni config location: %s\n",
+					filepath.Join(baseDir, OmniRelativePath), filepath.Join(xdg.ConfigHome, OmniRelativePath))
+
+				// Talos home directory is not writable, but we can still read the config from XDG config directory
+				return xdg.ConfigFile(OmniRelativePath)
+			}
+
+			return ensurePath(filepath.Join(baseDir, omniConfigFolder))
+		}
+
+		return "", errors.New("unknown access type")
+	}
+
+	if access == READ && fileutils.FileExists(filepath.Join(xdg.ConfigHome, OmniRelativePath)) {
+		return xdg.ConfigFile(OmniRelativePath)
+	}
+
+	return "", err
+}
+
+func ensurePath(path string) (string, error) {
+	err := os.MkdirAll(path, os.ModeDir|0o700)
```

Comment:
I can suggest doing inline error returns on calls which return a single value which is an error. There are multiple.